### PR TITLE
Add support for number literals

### PR DESCRIPTION
**Pull request by saswatamcode**
This PR adds support for number literals in the new engine (for both instant and range queries) as mentioned in #15.
Also adds `vector` function to make it easier to work with in query expressions.

### CODE REVIEW

File: physicalplan/scan/literal_selector.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	"math"
+	"time"
+
+	"github.com/thanos-community/promql-engine/physicalplan/model"
+
+	"github.com/prometheus/prometheus/model/labels"
+)
+
+// literalSelector returns []model.StepVector with same sample value across time range.
+type literalSelector struct {
```
Comment (yeya24) [inline physicalplan/scan/literal_selector.go:17]: Probably rename it to `numberLiteralSelector` as it is not string literal?

File: physicalplan/scan/literal_selector.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+	return o.vectorPool
+}
+
+func (o *numberLiteralSelector) Next(ctx context.Context) ([]model.StepVector, error) {
+	if o.currentStep > o.maxt {
+		return nil, nil
+	}
+
+	totalSteps := int64(1)
+	if o.step != 0 {
```
Comment (fpetkovski) [inline physicalplan/scan/literal_selector.go:62]: I wonder if we can extract this logic somewhere so we don't have to repeat it. My initial idea was to add a `NumSteps()` method to the `query.Options`, and inject an instance in relevant scanners.

Comment (saswatamcode) [reply physicalplan/scan/literal_selector.go:62]: Yup! That seems good to me too, but seems like we don't use `query.Options` anywhere else. Let me try to refactor this a bit in the next PR! 🙂

File: physicalplan/plan.go [hunk 3]
```diff
@@ -66,9 +66,18 @@ func newOperator(expr parser.Expr, storage storage.Queryable, mint, maxt time.Ti
 				operators = append(operators, exchange.NewConcurrent(selector, 2))
 			}
 			return exchange.NewCoalesce(model.NewVectorPool(stepsBatch), operators...), nil
+		case *parser.NumberLiteral:
```
Comment (fpetkovski) [inline physicalplan/plan.go:69]: Do we have an example of when this can happen?

File: physicalplan/scan/literal_selector.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+		totalSteps = (o.maxt-o.mint)/o.step + 1
+	}
+
+	numSteps := int(math.Min(float64(o.stepsBatch), float64(totalSteps)))
+
+	vectors := o.vectorPool.GetVectorBatch()
+	ts := o.currentStep
+
```
Comment (fpetkovski) [inline physicalplan/scan/literal_selector.go:70]: We can organize the variables a bit:
```suggestion
	vectors := o.vectorPool.GetVectorBatch()
	numSteps := int(math.Min(float64(o.stepsBatch), float64(totalSteps)))
	ts := o.currentStep
```

File: physicalplan/scan/literal_selector.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+		result := promql.Sample{
+			Point: promql.Point{
+				T: ts,
+				V: o.val,
+			},
+		}
+
+		if o.call != nil {
+			result = o.call(labels.New(), []promql.Point{result.Point}, time.UnixMilli(ts))
```
Comment (fpetkovski) [inline physicalplan/scan/literal_selector.go:84]: Can we use the existing labels initialized for the Series call instead of allocating a new slice?

**Review by fpetkovski — COMMENTED** [review]
Just a few more nits :)

File: physicalplan/plan.go (hunk 3 above)
Comment (saswatamcode) [reply physicalplan/plan.go:69]: Yup, the example can be anything like `vector(1)` or PromQL functions which have scalar input. This is also present as a testcase. 🙂

File: physicalplan/scan/function.go [hunk 6]
```diff
@@ -68,6 +68,13 @@ func NewFunctionCall(f *parser.Function, selectRange time.Duration) (FunctionCal
 				Metric: labels,
 			}
 		}, nil
+	case "vector":
+		return func(_ labels.Labels, points []promql.Point, stepTime time.Time) promql.Sample {
```
Comment (fpetkovski) [inline physicalplan/scan/function.go:72]: Should the labels be reused instead of allocating new ones?

Comment (saswatamcode) [reply physicalplan/scan/function.go:72]: Nice catch! Missed updating this.

**Review by fpetkovski — APPROVED** [review]
Thanks, this looks good 👍

**Review by yeya24 — APPROVED** [review]
Looks good!

**Pull request merged**